File: src/NerdBank.GitVersioning/CloudBuildServices/Jenkins.cs
// Copyright (c) .NET Foundation and Contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.Text;

namespace Nerdbank.GitVersioning.CloudBuildServices;

/// <summary>
/// Cloud build handling for Jenkins.
/// </summary>
/// <remarks>
/// The Jenkins-specific properties referenced here are <see href="https://wiki.jenkins-ci.org/display/JENKINS/Git+Plugin#GitPlugin-Environmentvariables">documented here</see>.
/// </remarks>
internal class Jenkins : ICloudBuild
{
    private static readonly Encoding UTF8NoBOM = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);

    /// <inheritdoc/>
    public string BuildingTag => null;

    /// <inheritdoc/>
    public bool IsPullRequest => false;

    /// <inheritdoc/>
    public string BuildingBranch => CloudBuild.ShouldStartWith(Branch, "refs/heads/");

    /// <inheritdoc/>
    public string GitCommitId => Environment.GetEnvironmentVariable("GIT_COMMIT");

    /// <inheritdoc/>
    public bool IsApplicable => !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("JENKINS_URL"));

    private static string Branch =>
        Environment.GetEnvironmentVariable("GIT_LOCAL_BRANCH")
            ?? Environment.GetEnvironmentVariable("GIT_BRANCH");

    /// <inheritdoc/>
    public IReadOnlyDictionary<string, string> SetCloudBuildNumber(string buildNumber, TextWriter stdout, TextWriter stderr)
    {
        WriteVersionFile(buildNumber);

        stdout.WriteLine($"## GIT_VERSION: {buildNumber}");

        return new Dictionary<string, string>
        {
            { "GIT_VERSION", buildNumber },
        };
    }

    /// <inheritdoc/>
    public IReadOnlyDictionary<string, string> SetCloudBuildVariable(string name, string value, TextWriter stdout, TextWriter stderr)
    {
        return new Dictionary<string, string>
        {
            { name, value },
        };
    }

    private static void WriteVersionFile(string buildNumber)
    {
        string workspacePath = Environment.GetEnvironmentVariable("WORKSPACE");

        if (string.IsNullOrEmpty(workspacePath))
        {
            return;
        }

        string versionFilePath = Path.Combine(workspacePath, "jenkins_build_number.txt");

        Utilities.FileOperationWithRetry(() => File.WriteAllText(versionFilePath, buildNumber, UTF8NoBOM));
    }
}

File: src/nerdbank-gitversioning-rs/src/cloud_build_services/jenkins.rs
// Copyright (c) .NET Foundation and Contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

//! [Jenkins Git plugin](https://plugins.jenkins.io/git/#plugin-content-environment-variables) support.

use std::fs;
use std::io::{self, Write};
use std::path::Path;

use crate::cloud::{CloudBuild, EnvironmentVariables, env, is_nonempty, should_start_with};

/// Jenkins cloud-build integration.
pub struct Jenkins;

impl CloudBuild for Jenkins {
    fn name(&self) -> &'static str {
        "Jenkins"
    }

    fn is_applicable(&self) -> bool {
        is_nonempty(env("JENKINS_URL").as_deref())
    }

    fn is_pull_request(&self) -> bool {
        false
    }

    fn building_branch(&self) -> Option<String> {
        should_start_with(
            env("GIT_LOCAL_BRANCH").or_else(|| env("GIT_BRANCH")),
            "refs/heads/",
        )
    }

    fn building_tag(&self) -> Option<String> {
        None
    }

    fn git_commit_id(&self) -> Option<String> {
        env("GIT_COMMIT")
    }

    fn set_cloud_build_number(
        &self,
        build_number: &str,
        stdout: &mut dyn Write,
        _stderr: &mut dyn Write,
    ) -> io::Result<EnvironmentVariables> {
        if let Some(workspace) = env("WORKSPACE").filter(|value| !value.is_empty()) {
            fs::write(
                Path::new(&workspace).join("jenkins_build_number.txt"),
                build_number.as_bytes(),
            )?;
        }
        writeln!(stdout, "## GIT_VERSION: {build_number}")?;
        Ok(EnvironmentVariables::from([(
            "GIT_VERSION".into(),
            build_number.into(),
        )]))
    }

    fn set_cloud_build_variable(
        &self,
        name: &str,
        value: &str,
        _stdout: &mut dyn Write,
        _stderr: &mut dyn Write,
    ) -> io::Result<EnvironmentVariables> {
        Ok(EnvironmentVariables::from([(name.into(), value.into())]))
    }
}
